src/org/zaproxy/zap/extension/ascanrulesAlpha/SQLInjectionMsSQL.java

```diff
@@ -0,0 +1,319 @@
+			//if the time was very slow (because JSP was being compiled on first call, for instance)
+			//then the rest of the time based logic will fail.  Lets double-check for that scenario by requesting the url again.
+			//If it comes back in a more reasonable time, we will use that time instead as our baseline.  If it come out in a slow fashion again,
+			//we will abort the check on this URL, since we will only spend lots of time trying request, when we will (very likely) not get positive results.
+			if (originalTimeUsed > sleep * 1000) {
+				long originalTimeStarted2 = System.currentTimeMillis();
+				try {
+					sendAndReceive(msgTimeBaseline, false); //do not follow redirects
+				}
+				catch (java.net.SocketTimeoutException e) {
+					//to be expected occasionally, if the base query was one that contains some parameters exploiting time based SQL injection?
+					if ( this.debugEnabled ) log.debug("Base Time Check 2 timed out on ["+msgTimeBaseline.getRequestHeader().getMethod()+"] URL ["+msgTimeBaseline.getRequestHeader().getURI().getURI()+"]");
+				} catch (SocketException ex) {
+					if ( this.debugEnabled ) log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + " when accessing: " + msgTimeBaseline.getRequestHeader().getURI().toString());
+					return; // No need to keep going
+				}
+				long originalTimeUsed2 = System.currentTimeMillis() - originalTimeStarted2;
+				if ( originalTimeUsed2 > sleep * 1000 ) {
+					//no better the second time around.  we need to bale out.
+					if ( this.debugEnabled ) log.debug("Both base time checks 1 and 2 for ["+msgTimeBaseline.getRequestHeader().getMethod()+"] URL ["+msgTimeBaseline.getRequestHeader().getURI().getURI()+"] are way too slow to be usable for the purposes of checking for time based SQL Injection checking.  We are aborting the check on this particular url.");
+					return;
+				} else {
+					//phew.  the second time came in within the limits. use the later timing details as the base time for the checks.
+					originalTimeUsed = originalTimeUsed2;
+					originalTimeStarted = originalTimeStarted2;
+				}
+			}
+			//end of timing baseline check
+
+			int countTimeBasedRequests = 0;
+
+			if ( this.debugEnabled ) log.debug("Scanning URL ["+ getBaseMsg().getRequestHeader().getMethod()+ "] ["+ getBaseMsg().getRequestHeader().getURI() + "], field ["+ paramName + "] with value ["+paramValue+"] for MsSQL Injection");
+
+			//Check for time based SQL Injection, using MsSQL specific syntax
+			for (int timeBasedSQLindex = 0;
+					timeBasedSQLindex < SQL_MSSQL_TIME_REPLACEMENTS.length && doTimeBased && countTimeBasedRequests < doTimeMaxRequests;
+					timeBasedSQLindex ++) {
+				HttpMessage msgAttack = getNewMsg();
+				String newTimeBasedInjectionValue =
+						SQL_MSSQL_TIME_REPLACEMENTS[timeBasedSQLindex].
+								replace (ORIG_VALUE_TOKEN, paramValue).
+								replace(SLEEP_TOKEN, getSleepToken(sleep));
```

Comment:
`getSleepToken()` return value is always the same, it could be done outside the loop.